File: examples/javascript/websocket-client.js
const BASE_URL = 'http://localhost:8080/api';
const WS_URL = 'ws://localhost:8080/ws';

async function login(username, password) {
  const res = await fetch(`${BASE_URL}/login`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name: username, pass: password }),
  });
  if (!res.ok) throw new Error(`Login failed: ${res.status}`);
  return res.json();
}

function connectWebSocket(token) {
  return new Promise((resolve, reject) => {
    const ws = new WebSocket(`${WS_URL}?token=${token}`);

    ws.onopen = () => {
      console.log('   WebSocket connected');
      resolve(ws);
    };

    ws.onmessage = (event) => {
      const data = JSON.parse(event.data);
      if (data.type === 'connected') {
        console.log(`   Connected as user ID: ${data.data.id}`);
      } else if (data.type === 'message') {
        const msg = data.data;
        console.log(`\n   New message received:`);
        console.log(`   - App ID: ${msg.appid}`);
        console.log(`   - Title: ${msg.title}`);
        console.log(`   - Message: ${msg.message}`);
        console.log(`   - Priority: ${msg.priority}`);
      }
    };

    ws.onerror = (err) => {
      console.error('   WebSocket error:', err);
      reject(err);
    };

    ws.onclose = (event) => {
      console.log(`   WebSocket closed: ${event.code} ${event.reason}`);
    };
  });
}

async function main() {
  console.log('=== Miotify WebSocket Example ===\n');

  console.log('1. Login');
  const loginRes = await login('admin', 'admin');
  console.log(`   Token received`);
  const token = loginRes.token;

  console.log('\n2. Connect WebSocket');
  const ws = await connectWebSocket(token);

  console.log('\n3. Listening for messages...');
  console.log('   (Send a message to your app to see it here)');
  console.log('   Press Ctrl+C to exit\n');
}

main().catch(console.error);

File: examples/python/websocket-client.py
import requests
import asyncio
import websockets
import json

BASE_URL = 'http://localhost:8080/api'
WS_URL = 'ws://localhost:8080/ws'

def login(username, password):
    res = requests.post(f'{BASE_URL}/login', json={'name': username, 'pass': password})
    res.raise_for_status()
    return res.json()

async def listen_messages(token):
    uri = WS_URL
    print('   Connecting to WebSocket...')
    # token 通过 Sec-WebSocket-Protocol 子协议传递（避免出现在 URL query 中被日志记录）
    async with websockets.connect(uri, subprotocols=['miotify', token]) as ws:
        print('   WebSocket connected\n')
        print('3. Listening for messages...')
        print('   (Send a message to your app to see it here)')
        print('   Press Ctrl+C to exit\n')
        
        async for message in ws:
            data = json.loads(message)
            if data['type'] == 'connected':
                print(f"   Connected as user ID: {data['data']['id']}")
            elif data['type'] == 'message':
                msg = data['data']
                print(f"\n   New message received:")
                print(f"   - App ID: {msg['appid']}")
                print(f"   - Title: {msg['title']}")
                print(f"   - Message: {msg['message']}")
                print(f"   - Priority: {msg['priority']}")

async def main():
    print('=== Miotify WebSocket Example ===\n')

    print('1. Login')
    login_res = login('admin', 'admin')
    print('   Token received')
    token = login_res['token']

    print('\n2. Connect WebSocket')
    await listen_messages(token)

if __name__ == '__main__':
    asyncio.run(main())
